### ISSUE TITLE
[RFC] NVIDIA CUDA MPS for colocated reward workers

### ISSUE THREAD

**Issue by CjhHa1** [Tencent-Hunyuan/UniRL#463]
**Status:** Draft
**Scope:** standalone Reward Service, Ray GPU placement, NVIDIA CUDA MPS

## Summary

This RFC proposes opt-in NVIDIA CUDA Multi-Process Service (MPS) support for
the standalone `unirl-reward-service`, initially targeting several small,
independent reward models on one physical GPU.

The proposed division of responsibilities is:

```text
Ray fractional num_gpus   -> placement: which actors share a physical GPU
CUDA MPS                   -> execution: allow kernels from those CUDA contexts to overlap
per-client MPS limits      -> guardrails: cap compute visibility and device allocations
```

This is a capacity and utilization feature, not a blanket performance claim.
Implementation should proceed only after an MPS/no-MPS benchmark on the same
physical GPU shows a useful workload-specific result.

## Motivation

The Reward Service already has the right high-level concurrency:

1. The FastAPI gateway buckets one request by reward name.
2. It dispatches each reward bucket concurrently.
3. Each reward runs in its own Ray `ScorerActor`, preserving dependency and
   failure isolation at the Python-process level.

Today the documented deployment model gives each actor exclusive GPU ownership.
That wastes cards for panels of small rewards such as PickScore, CLIP, and
HPSv2, where one model may leave much of an H20 idle.

The source is already close to a proof of concept:

- `RewardModelCfg.num_gpus` is a `float`.
- `WorkerGroup._actor_options()` forwards it directly to Ray.
- Ray fractional GPU resources can therefore place multiple actors on the same
  physical card.

However, Ray fractions are logical scheduling resources only. They do not make
independent CUDA contexts execute concurrently and do not enforce GPU-memory
limits. Without MPS, co-resident reward actors can still time-slice, over-allocate
memory, and OOM one another.

## Goals

- Run multiple independent, single-GPU reward actors on one physical NVIDIA GPU.
- Preserve one process/runtime environment per reward.
- Keep MPS strictly opt-in; existing configurations behave unchanged.
- Use Ray fractional GPUs for placement and MPS for cross-context execution.
- Configure per-reward active-thread and device-memory allocation limits before
  the first CUDA context is created.
- Fail fast on unsupported combinations.
- Provide benchmark evidence for throughput, latency, memory, and GPU efficiency.
- Shut clients and the MPS daemon down safely without stale processes or pipes.

## Non-goals

- Train + rollout + reward sharing in this first RFC.
- Multiple FSDP/DDP/NCCL ranks on the same physical GPU.
- Shared-MPS placement for TP>1 reward models.
- Making MPS the default.
- Requiring CUDA 13.4 / MPS v3.
- Automatically changing GPU compute mode.
- Providing MIG-grade memory, bandwidth, QoS, or fault isolation.
- Deduplicating weights, KV caches, or CUDA graph pools across processes.

## Why MPS may help

CUDA MPS is useful when individual processes do not saturate the GPU. It allows
kernels and memory copies from different CUDA contexts to overlap and reduces
context-switching overhead.

The strongest initial candidate is a panel of small Transformer-based rewards
running with small or bursty batches. A single request already fans out to these
rewards concurrently, so the service can expose real cross-process overlap
without changing its API or request semantics.

MPS is unlikely to help when one scorer already saturates SMs or memory
bandwidth. Multiple vLLM/SGLang instances may also lose to a consolidated,
continuously batched process because every instance duplicates weights, KV
cache, allocator state, and CUDA graph memory. Those cases must be benchmarked,
not assumed.

References:

- [NVIDIA: When to Use MPS](https://docs.nvidia.com/deploy/mps/latest/when-to-use-mps.html)
- [NVIDIA: MPS Architecture](https://docs.nvidia.com/deploy/mps/latest/architecture.html)
- [Ray: Accelerator Support](https://docs.ray.io/en/latest/ray-core/scheduling/accelerators.html)

### 1. Benchmark gate

Add a reproducible four-way comparison:

1. Each reward on a dedicated GPU.
2. Multiple fractional Ray actors on one GPU without MPS.
3. The same fractional placement with MPS.
4. A consolidated/continuously batched process where dependencies permit.

Use representative production shapes and sweep request batch size and
concurrency. Record:

- panel items/s and rewards/GPU-hour;
- per-reward and panel p50/p95/p99 latency;
- peak and steady device memory;
- SM, Tensor Core, and DRAM activity;
- OOMs, score failures, MPS `FAULT` transitions, and restarts;
- numerical score parity.

MPS proceeds as a throughput optimization only if repeated same-card runs beat
the no-MPS result beyond measurement noise without harming stability. If it
only reduces card count, it can still ship as an explicitly documented capacity
mode.

### 2. Node-level MPS lifecycle

Start MPS once per node before Ray creates any CUDA actors. Use short,
job/UID-scoped pipe and log directories and provide start/status/stop commands.

Legacy MPS v2 should be the compatibility baseline. MPS v3 can be detected and
supported later, but must not be required by the initial implementation.

The launch layer owns the daemon; actor code only receives client environment
variables. Existing externally managed MPS daemons should also be usable without
UniRL trying to replace them.

### 3. Per-reward client configuration

Extend the typed reward config with an explicit nested MPS client block, for
example:

```yaml
rewards:
  - name: pickscore
    scorer: pickscore
    num_gpus: 0.5
    mps:
      active_thread_percentage: 50
      device_memory_limit: 12GiB
```

The exact schema should have one owner for each default and should not duplicate
the scorer's existing memory settings.

`WorkerGroup` injects the validated MPS environment through Ray `runtime_env`
before actor startup. `ScorerActor._pin_cuda_to_ray_gpus()` remains: each actor
should still see exactly the one GPU assigned by Ray, even when another actor is
assigned the same card.

Validation should reject at least:

- shared-MPS profiles with TP>1;
- non-finite or out-of-range fractional GPU requests;
- invalid active-thread percentages or memory sizes;
- MPS client settings when the node-level runtime is unavailable;
- incompatible vLLM memory settings where the effective budgets overlap.

The MPS device-memory limit is an allocation cap, not a physical partition.
Clients still share memory bandwidth, cache, and a GPU fault domain.

### 4. Graceful shutdown and fault semantics

Raw termination of an active MPS client can leave other clients or the MPS
server in an undefined state. Reward actors should first stop admission, drain,
synchronize outstanding CUDA work, release the scorer, and only then be killed.
The daemon stops after all clients have exited.

The current per-reward error-isolation guarantee must be qualified: an ordinary
Python/scorer failure remains isolated, but a fatal GPU fault can affect every
MPS client sharing that GPU. Health output and documentation should identify
the shared fault domain and provide a whole-group restart procedure.

### PR 1: Benchmark harness and baseline report

- Extend `unirl-reward-service/scripts/bench_concurrent.py` for the four-way
  comparison and machine-readable results.
- Add environment/version capture and GPU/MPS observations.
- Publish H20 results for a small reward panel.
- No production scheduling behavior changes.

### PR 2: MPS runtime lifecycle

- Add reusable start/status/stop helpers.
- Integrate opt-in startup before Ray on single- and multi-node reward-service
  launch paths.
- Add capability checks, scoped pipe/log directories, and safe cleanup.
- Support an externally managed daemon.

### PR 3: Fractional reward placement and client budgets

- Add the typed nested MPS client config.
- Inject client environment before CUDA initialization.
- Add TP/fraction/memory compatibility validation.
- Preserve the non-MPS path byte-for-byte where practical.
- Add CPU config tests and a GPU placement smoke test.

### PR 4: Reference recipes and documentation

- Add an H20 recipe for PickScore + CLIP/HPSv2 sharing one GPU.
- Add a mixed recipe where small rewards share a card and a large judge keeps
  an exclusive card.
- Update the README and architecture document, which currently describe GPU
  ownership as always exclusive.
- Document capacity mode versus measured throughput mode.

### PR 5: Reliability and backend qualification

- Add long-running soak tests and MPS `FAULT`/OOM/cleanup scenarios.
- Qualify pure PyTorch/Transformers scorers first.
- Evaluate vLLM TP=1 separately with eager/CUDA-graph and memory-budget sweeps.
- Keep unqualified backends rejected at configuration time.

Each PR should be independently reviewable and preserve existing default
behavior.

## Acceptance criteria

- MPS-disabled configurations retain their existing placement and results.
- Two or more configured reward actors are verified to share one physical GPU
  through Ray while connecting to the intended MPS server.
- Scores match dedicated-GPU baselines within each scorer's established
  numerical tolerance.
- Per-client allocation limits fail locally with actionable diagnostics instead
  of silently exhausting the whole card.
- Unsupported TP/NCCL combinations fail during configuration or startup.
- Graceful shutdown leaves no live MPS clients or stale job-scoped runtime.
- A 12–24 hour H20 soak completes without OOM, MPS `FAULT`, stuck Ray actors, or
  score corruption.
- The benchmark report includes no performance claim based only on average GPU
  utilization.

## Risks

- One heavy reward can consume shared SM or memory bandwidth and worsen panel
  tail latency.
- CUDA graph pools and framework allocations can invalidate apparently safe
  model-weight budgets.
- `nvidia-smi`/NVML may attribute activity to the MPS server, making per-client
  observability harder.
- A fatal GPU fault expands from one actor to every client sharing that card.
- Driver/CUDA/framework combinations may behave differently; support must be
  version-qualified.

## Open questions

1. Is PickScore + CLIP + HPSv2 the right first H20 panel, or should the pilot use
   the most common production reward set?
2. Should PR 1 reuse the benchmark result schema proposed in #414 if available,
   or emit a local schema without blocking on that RFC?
3. Should UniRL own the MPS daemon in production pods, or should the platform
   inject an already-running daemon and its pipe directory?
4. Should the first MVP stop at pure PyTorch/Transformers scorers and leave all
   vLLM rewards for PR 5?
5. What workload-specific latency and cost thresholds should graduate a recipe
   from experimental to supported?

**Comment by jyizheng** [comment]
I'd like to take PR 1 (benchmark harness and baseline report).

Files: `unirl-reward-service/scripts/bench_concurrent.py` (four-way comparison, machine-readable results, env/version capture) plus a README note. No scheduling behavior changes.

One caveat: I have B200/H100 but no H20. I'll publish B200/H100 numbers labeled as such; the H20 report would need a maintainer run of the same harness.

Domain reviewer: not needed.

**Comment by jyizheng** [comment]
Just noticed @Zcchill is already assigned — happy to coordinate instead of taking PR 1 outright. I can run the four-way benchmark on B200/H100 once the harness shape is agreed, or review the MPS lifecycle pieces.

**Comment by CjhHa1** [comment]
> Just noticed [@Zcchill](https://github.com/Zcchill) is already assigned — happy to coordinate instead of taking PR 1 outright. I can run the four-way benchmark on B200/H100 once the harness shape is agreed, or review the MPS lifecycle pieces.

Hi，would you mind contact me via wechat or any else? Mine is CjhHa1

BTW, you are welcome to join and benchmarking any configure you want
